// appium-youiengine-driver/build/lib/basedevice.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _logger = _interopRequireDefault(require("./logger"));

class BaseDevice {
  constructor() {}

  closeApp() {
    _logger.default.errorAndThrow(`closeApp is not supported on this platform`);
  }

  endSession() {
    _logger.default.errorAndThrow(`endSession is not supported on this platform`);
  }

  installApp(appPath) {
    _logger.default.errorAndThrow(`installApp is not supported on this platform`);
  }

  isAppInstalled(channelId) {
    _logger.default.errorAndThrow(`isAppInstalled is not supported on this platform`);
  }

  launchApp() {
    _logger.default.errorAndThrow(`launchApp is not supported on this platform`);
  }

  removeApp(channelId) {
    _logger.default.errorAndThrow(`removeApp is not supported on this platform`);
  }

  startSession(caps) {
    _logger.default.errorAndThrow(`startSession is not supported on this platform`);
  }

}

var _default = BaseDevice;
exports.default = _default;require('source-map-support').install();
