// appium-ios-simulator/build/lib/tcc-db.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _path = _interopRequireDefault(require("path"));

var _appiumSupport = require("appium-support");

var _utils = require("./utils");

class TCCDB {
  constructor(xcodeVersion, sharedResourcesDir) {
    this.xcodeVersion = xcodeVersion;
    this.sharedResourcesDir = sharedResourcesDir;
  }

  async getDB() {
    if (this.db) {
      return this.db;
    }

    const tccPath = _path.default.resolve(this.sharedResourcesDir, 'Library', 'TCC');

    if (!(await _appiumSupport.fs.exists(tccPath))) {
      await (0, _appiumSupport.mkdirp)(tccPath);
    }

    this.db = _path.default.resolve(tccPath, 'TCC.db');
    await (0, _utils.execSQLiteQuery)(this.db, `CREATE TABLE IF NOT EXISTS access (
      service TEXT NOT NULL DEFAULT '',
      client TEXT NOT NULL DEFAULT '',
      client_type INTEGER,
      allowed INTEGER,
      prompt_count INTEGER,
      csreq BLOB NOT NULL DEFAULT '',
      policy_ID INTEGER,
      PRIMARY KEY(service, client, client_type)
    );`);
    return this.db;
  }

  async execQuery(query, ...params) {
    return await (0, _utils.execSQLiteQuery)((await this.getDB()), query, ...params);
  }

}

var _default = TCCDB;
exports.default = _default;require('source-map-support').install();
