// appium-xcode/build/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.getClangVersion = exports.getMaxTVOSSDKWithoutRetry = exports.getMaxTVOSSDK = exports.getCommandLineToolsVersion = exports.getInstrumentsPath = exports.clearInternalCache = exports.getConnectedDevices = exports.getMaxIOSSDKWithoutRetry = exports.getAutomationTraceTemplatePathWithoutRetry = exports.getMaxIOSSDK = exports.getAutomationTraceTemplatePath = exports.getVersion = exports.getPath = void 0;

require("source-map-support/register");

var xcode = _interopRequireWildcard(require("./lib/xcode"));

const {
  getPath,
  getVersion,
  getAutomationTraceTemplatePath,
  getMaxIOSSDK,
  getAutomationTraceTemplatePathWithoutRetry,
  getMaxIOSSDKWithoutRetry,
  getConnectedDevices,
  clearInternalCache,
  getInstrumentsPath,
  getCommandLineToolsVersion,
  getMaxTVOSSDK,
  getMaxTVOSSDKWithoutRetry,
  getClangVersion
} = xcode;
exports.getClangVersion = getClangVersion;
exports.getMaxTVOSSDKWithoutRetry = getMaxTVOSSDKWithoutRetry;
exports.getMaxTVOSSDK = getMaxTVOSSDK;
exports.getCommandLineToolsVersion = getCommandLineToolsVersion;
exports.getInstrumentsPath = getInstrumentsPath;
exports.clearInternalCache = clearInternalCache;
exports.getConnectedDevices = getConnectedDevices;
exports.getMaxIOSSDKWithoutRetry = getMaxIOSSDKWithoutRetry;
exports.getAutomationTraceTemplatePathWithoutRetry = getAutomationTraceTemplatePathWithoutRetry;
exports.getMaxIOSSDK = getMaxIOSSDK;
exports.getAutomationTraceTemplatePath = getAutomationTraceTemplatePath;
exports.getVersion = getVersion;
exports.getPath = getPath;
var _default = xcode;
exports.default = _default;require('source-map-support').install();
