// appium-xcuitest-driver/build/lib/wda/iproxy.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.iProxy = void 0;

require("source-map-support/register");

var _logger = _interopRequireDefault(require("../logger"));

var _utils = require("./utils");

var _bluebird = _interopRequireDefault(require("bluebird"));

var _appiumSupport = require("appium-support");

var _teen_process = require("teen_process");

var _portscanner = require("portscanner");

var _asyncbox = require("asyncbox");

const IPROXY_STARTUP_TIMEOUT = 5000;

const iproxyLog = _appiumSupport.logger.getLogger('iProxy');

class iProxy {
  constructor(udid, localport, deviceport, detached = true) {
    this.expectIProxyErrors = true;
    this.localport = parseInt(localport, 10);
    this.deviceport = parseInt(deviceport, 10);
    this.udid = udid;
    this.iproxy = new _teen_process.SubProcess('iproxy', [localport, deviceport, udid], {
      detached,
      stdio: ['ignore', 'pipe', 'pipe']
    });
  }

  async start() {
    _logger.default.debug(`Starting iproxy to forward traffic from local port ${this.localport} ` + `to device port ${this.deviceport} over USB for the device ${this.udid}`);

    this.expectIProxyErrors = true;
    return await new _bluebird.default((resolve, reject) => {
      this.iproxy.on('exit', code => {
        _logger.default.debug(`iproxy exited with code '${code}'`);

        if (code) {
          return reject(new Error(`iproxy exited with code '${code}'`));
        }
      });
      this.iproxy.on('output', (stdout, stderr) => {
        if (this.expectIProxyErrors) {
          return;
        }

        let out = stdout || stderr;

        for (let line of out.split('\n')) {
          if (!line.length) {
            continue;
          }

          if (line.includes('Resource temporarily unavailable')) {
            _logger.default.debug('Connection to WDA timed out');
          } else {
            iproxyLog.debug(line);
          }
        }
      });
      return (async () => {
        try {
          if ((await (0, _portscanner.checkPortStatus)(this.localport, '127.0.0.1')) === 'open') {
            throw new Error(`The port #${this.localport} is occupied by an other app. ` + `You can customize its value by setting the 'wdaLocalPort' capability.`);
          }

          await this.iproxy.start(0);

          try {
            await (0, _asyncbox.waitForCondition)(async () => {
              try {
                return (await (0, _portscanner.checkPortStatus)(this.localport, '127.0.0.1')) === 'open';
              } catch (ign) {
                return false;
              }
            }, {
              waitMs: IPROXY_STARTUP_TIMEOUT,
              intervalMs: 300
            });

            _logger.default.debug(`iProxy is running and is listening on port #${this.localport}`);
          } catch (e) {
            _logger.default.warn(`The local port ${this.localport} is still closed after ${IPROXY_STARTUP_TIMEOUT}ms. ` + `Continuing anyway`);
          }

          resolve();
        } catch (err) {
          _logger.default.error(`Error starting iproxy: '${err.message}'`);

          reject(new Error('Unable to start iproxy. Make sure libusbmuxd is installed and ' + 'PATH contains the folder, where the binary is located.'));
        }
      })();
    });
  }

  async quit() {
    await (0, _utils.killProcess)('iproxy', this.iproxy);
    this.expectIProxyErrors = true;
  }

}

exports.iProxy = iProxy;
var _default = iProxy;
exports.default = _default;require('source-map-support').install();
