// appium-ios-driver/build/lib/uiauto/dynamic-bootstrap.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.prepareBootstrap = prepareBootstrap;
exports.getEnv = getEnv;

require("source-map-support/register");

var _path = _interopRequireDefault(require("path"));

var _lodash = _interopRequireDefault(require("lodash"));

var _crypto = _interopRequireDefault(require("crypto"));

var _appiumSupport = require("appium-support");

var _buildScript = _interopRequireDefault(require("./build-script"));

var _logger = _interopRequireDefault(require("./logger"));

var _uiautoClient = require("./uiauto-client");

let BOOTSTRAP_JS_PATH = _path.default.resolve(__dirname, '..', '..', '..', 'uiauto', 'bootstrap.js');

let COMMAND_PROXY_CLIENT_PATH = _path.default.resolve(__dirname, 'bin', 'command-proxy-client.js');

if (!__dirname.match(/build\/lib\/uiauto$/)) {
  BOOTSTRAP_JS_PATH = _path.default.resolve(__dirname, '..', 'uiauto', 'bootstrap.js');
  COMMAND_PROXY_CLIENT_PATH = _path.default.resolve(__dirname, 'bin', 'command-proxy-client.js');
}

function getEnv(opts = {}) {
  return {
    nodePath: process.execPath,
    commandProxyClientPath: COMMAND_PROXY_CLIENT_PATH,
    instrumentsSock: opts.sock || _uiautoClient.DEFAULT_INSTRUMENTS_SOCKET,
    interKeyDelay: opts.interKeyDelay || null,
    justLoopInfinitely: opts.justLoopInfinitely,
    autoAcceptAlerts: opts.autoAcceptAlerts,
    autoDismissAlerts: opts.autoDismissAlerts,
    sendKeyStrategy: opts.sendKeyStrategy,
    initialLocation: opts.initialLocation
  };
}

async function buildCode(opts) {
  if (opts.code) return opts.code;
  let env = getEnv(opts);

  _logger.default.debug(`Dynamic env: ${JSON.stringify(env)}`);

  let bootstrapJs = BOOTSTRAP_JS_PATH;
  let imports = opts.imports && opts.imports.pre ? opts.imports.pre : [];
  let bootstrapCode = await (0, _buildScript.default)(bootstrapJs, imports);
  let lines = [];
  lines.push('// This file is automatically generated. Do not manually modify!');
  lines.push('');
  lines.push(bootstrapCode);
  lines.push('');
  lines.push('bootstrap({');

  for (let [key, value] of _lodash.default.toPairs(env)) {
    if (!_lodash.default.isUndefined(value)) {
      let quote = _lodash.default.isString(value) ? '\"' : '';
      lines.push(`  "${key}": ${quote}${value}${quote},`);
    }
  }

  lines[lines.length - 1] = lines[lines.length - 1].replace(/,$/, '');
  lines.push('});');
  return lines.join('\r\n');
}

function computeHash(code) {
  return _crypto.default.createHash('md5').update(code).digest('hex').substring(0, 16);
}

function getDynamicBootstrapDir(opts = {}) {
  let dynamicBootstrapDir;

  if (process.env.APPIUM_BOOTSTRAP_DIR) {
    dynamicBootstrapDir = process.env.APPIUM_BOOTSTRAP_DIR;
  } else if (process.env.HOME) {
    dynamicBootstrapDir = _path.default.resolve(process.env.HOME, 'Library/Application Support/appium/bootstrap');
  } else {
    dynamicBootstrapDir = _path.default.resolve(opts.tmpDir || '/tmp', 'appium/bootstrap');
  }

  return dynamicBootstrapDir;
}

async function writeDynamicBootstrapIfNecessary(dynamicBootstrapDir, dynamicBootstrapPath, code, hash) {
  await (0, _appiumSupport.mkdirp)(dynamicBootstrapDir);
  let codeIsGood = true;

  try {
    let existingCode = await _appiumSupport.fs.readFile(dynamicBootstrapPath);
    codeIsGood = computeHash(existingCode) === hash;
  } catch (err) {
    codeIsGood = false;
  }

  if (codeIsGood) {
    _logger.default.debug(`Reusing dynamic bootstrap: ${dynamicBootstrapPath}`);
  } else {
    _logger.default.debug(`Creating or overwriting dynamic bootstrap: ${dynamicBootstrapPath}`);

    await _appiumSupport.fs.writeFile(dynamicBootstrapPath, code, {
      flag: 'w+'
    });
  }
}

async function prepareBootstrap(opts = {}) {
  _logger.default.debug('Preparing bootstrap code');

  let dynamicBootstrapDir = getDynamicBootstrapDir(opts);

  _logger.default.debug(`Dynamic bootstrap dir: ${dynamicBootstrapDir}`);

  let code = await buildCode(opts);
  let hash = computeHash(code);

  let dynamicBootstrapPath = _path.default.resolve(dynamicBootstrapDir, `bootstrap-${hash}.js`);

  _logger.default.debug(`Dynamic bootstrap code: ${code.split('\n')[0]}...`);

  _logger.default.debug(`Dynamic bootstrap path: ${dynamicBootstrapPath}`);

  await writeDynamicBootstrapIfNecessary(dynamicBootstrapDir, dynamicBootstrapPath, code, hash);
  return dynamicBootstrapPath;
}require('source-map-support').install();
