// appium-ios-driver/build/lib/uiauto/uiauto-response.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _logger = _interopRequireDefault(require("./logger"));

var _lodash = _interopRequireDefault(require("lodash"));

const MESSAGE_TYPES = ['error', 'no data', 'regular', 'chunk', 'last chunk'];
const UNKNOWN_ERROR = {
  status: 13,
  value: 'Error parsing socket data from instruments'
};

class UIAutoResponse {
  constructor() {
    this.bufferedData = '';
    this.resultBuffer = '';
  }

  resetBuffer() {
    this.bufferedData = '';
  }

  addData(data) {
    this.bufferedData += data;
  }

  finalizeData() {
    let data = this.bufferedData;
    this.bufferedData = '';
    let parsedData;

    try {
      parsedData = {
        type: MESSAGE_TYPES[parseInt(data[0], 10)]
      };

      if (parsedData.type !== 'no data') {
        parsedData.result = data.substring(2);
      }
    } catch (err) {
      _logger.default.error(`Could not parse data from socket: ${err}`);

      _logger.default.error(data);

      parsedData = {
        type: 'error',
        error: UNKNOWN_ERROR
      };
    }

    return parsedData;
  }

  getResult() {
    let data = this.finalizeData();

    if (!_lodash.default.isUndefined(data.result) && data.result !== false) {
      if (data.result) {
        _logger.default.debug(`Got result from instruments: ${data.result.slice(0, 300)}`);
      } else {
        _logger.default.debug('Got null result from instruments');
      }

      if (data.type && data.type.indexOf('chunk') !== -1) {
        this.resultBuffer += data.result;

        _logger.default.debug(`Got chunk data, current resultBuffer length: ${this.resultBuffer.length}`);

        if (data.type === 'last chunk') {
          _logger.default.debug(`This is the last data final length: ${this.resultBuffer.length}`);

          let result;

          try {
            result = JSON.parse(this.resultBuffer);
          } catch (err) {
            _logger.default.error(`Could not parse result buffer: ${err}`);

            result = UNKNOWN_ERROR;
          }

          this.resultBuffer = '';
          return result;
        } else {
          _logger.default.debug('Not the last chunk, trying to get more');

          return {
            needsMoreData: true
          };
        }
      } else {
        let result;

        try {
          result = JSON.parse(data.result);
        } catch (err) {
          _logger.default.error(`Could not parse result buffer: ${err}`);

          result = UNKNOWN_ERROR;
        }

        return result;
      }
    } else {
      return null;
    }
  }

}

var _default = UIAutoResponse;
exports.default = _default;require('source-map-support').install();
