// appium-ios-driver/build/lib/uiauto/uiauto-client.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DEFAULT_INSTRUMENTS_SOCKET = exports.UIAutoClient = void 0;

require("source-map-support/register");

var _uiautoResponse = _interopRequireDefault(require("./uiauto-response"));

var _logger = _interopRequireDefault(require("./logger"));

var _through = _interopRequireDefault(require("through"));

var _net = _interopRequireDefault(require("net"));

var _appiumSupport = require("appium-support");

var _path = _interopRequireDefault(require("path"));

var _bluebird = _interopRequireDefault(require("bluebird"));

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumBaseDriver = require("appium-base-driver");

const MORE_COMMAND = '#more';
const DEFAULT_INSTRUMENTS_SOCKET = '/tmp/instruments_sock';
exports.DEFAULT_INSTRUMENTS_SOCKET = DEFAULT_INSTRUMENTS_SOCKET;

class UIAutoClient {
  constructor(sock = '/tmp/instruments_sock') {
    this.curCommand = null;
    this.onReceiveCommand = null;
    this.commandQueue = [];
    this.sock = sock;
    this.socketServer = null;
    this.hasConnected = false;
    this.currentSocket = null;
  }

  async sendCommand(cmd) {
    let cmdPromise = new _bluebird.default((resolve, reject) => {
      let cb = result => {
        if (result.status === 0) {
          resolve(result.value);
        } else if (result.status) {
          let jsonwpError = (0, _appiumBaseDriver.errorFromCode)(result.status, result.value);
          reject(jsonwpError);
        } else {
          reject(new Error(result.value));
        }
      };

      this.commandQueue.push({
        cmd,
        cb
      });

      if (_lodash.default.isFunction(this.onReceiveCommand)) {
        this.onReceiveCommand();
      }
    });
    return await cmdPromise;
  }

  async start() {
    let connectedPromise = new _bluebird.default(resolve => {
      let response = new _uiautoResponse.default();
      this.socketServer = _net.default.createServer({
        allowHalfOpen: true
      }, conn => {
        if (!this.hasConnected) {
          this.hasConnected = true;

          _logger.default.info('Instruments is ready to receive commands');

          resolve(true);
        }

        conn.setEncoding('utf8');
        this.currentSocket = conn;
        conn.on('close', () => {
          this.currentSocket = null;
        });
        conn.pipe((0, _through.default)(data => {
          _logger.default.debug(`Socket data received (${data.length} bytes)`);

          response.addData(data);
        }));
        conn.on('end', () => {
          if (this.curCommand) {
            let result = response.getResult();

            if (result && !result.needsMoreData) {
              this.curCommand.cb(result);
              this.curCommand = null;
            } else {
              if (result) {
                _logger.default.debug('Not the last chunk, trying to get more');
              } else {
                _logger.default.debug('No result received. Continuing to try to get more');
              }

              this.commandQueue.unshift({
                cmd: MORE_COMMAND,
                cb: this.curCommand.cb
              });
            }
          } else {
            _logger.default.debug('Got a result when we were not expecting one! Ignoring it');

            response.resetBuffer();
          }

          let onReceiveCommand = () => {
            this.onReceiveCommand = null;
            this.curCommand = this.commandQueue.shift();

            _logger.default.debug(`Sending command to instruments: ${this.curCommand.cmd}`);

            conn.write(JSON.stringify({
              cmd: this.curCommand.cmd
            }));
            conn.end();
          };

          if (this.commandQueue.length) {
            onReceiveCommand();
          } else {
            this.onReceiveCommand = onReceiveCommand;
          }
        });
      });
      this.socketServer.on('close', function () {
        _logger.default.debug('Instruments socket server was closed');
      });
    });
    await _appiumSupport.fs.rimraf(this.sock);
    await (0, _appiumSupport.mkdirp)(_path.default.dirname(this.sock));
    this.socketServer.listen(this.sock);

    _logger.default.debug(`Instruments socket server started at ${this.sock}`);

    return await connectedPromise;
  }

  async shutdown() {
    this.curCommand = null;
    this.onReceiveCommand = null;

    if (this.currentSocket) {
      _logger.default.debug('Destroying instruments client socket.');

      this.currentSocket.end();
      this.currentSocket.destroy();
      this.currentSocket = null;
    }

    if (this.socketServer) {
      _logger.default.debug('Closing socket server.');

      await _bluebird.default.promisify(this.socketServer.close, {
        context: this.socketServer
      })();
      this.socketServer = null;
    }
  }

  async safeShutdown() {
    _logger.default.debug('Shutting down command proxy and ignoring any errors');

    try {
      await this.shutdown();
    } catch (err) {
      _logger.default.debug(`Ignoring error: ${err}`);
    }
  }

}

exports.UIAutoClient = UIAutoClient;
var _default = UIAutoClient;
exports.default = _default;require('source-map-support').install();
