// appium-ios-driver/build/lib/instruments/streams.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.outputStream = outputStream;
exports.errorStream = errorStream;
exports.webSocketAlertStream = webSocketAlertStream;
exports.dumpStream = dumpStream;

require("source-map-support/register");

var _logger = _interopRequireDefault(require("./logger"));

var _through = _interopRequireDefault(require("through"));

function clearBufferChars(output) {
  let re = /(\n|^)\*+\n?/g;
  output = output.toString();
  output = output.replace(re, '');
  return output;
}

function outputStream() {
  return (0, _through.default)(function outputStreamHandler(output) {
    output = clearBufferChars(output);

    if (output !== '') {
      output = output.replace(/\n$/m, '');
      output = output.replace(/\n/m, '\n       ');
      output = `[INST] ${output}`;
      output = output.green;

      _logger.default.debug(output);
    }

    this.queue(output);
  });
}

function errorStream() {
  return (0, _through.default)(function errorStreamHandler(output) {
    output = output.replace(/\n$/m, '');
    output = `[INST STDERR] ${output}`;
    output = output.yellow;

    _logger.default.debug(output);

    this.queue(output);
  });
}

function webSocketAlertStream(webSocket) {
  return (0, _through.default)(function webSocketAlertStreamHandler(output) {
    if (webSocket) {
      let re = /Call to onAlert returned 'YES'/;

      if (re.test(output)) {
        _logger.default.debug('Emiting alert message...');

        webSocket.sockets.emit('alert', {
          message: output
        });
      }
    }

    this.queue(output);
  });
}

function dumpStream() {
  return (0, _through.default)(function nullStreamHandler() {});
}require('source-map-support').install();
