// appium-ios-driver/build/lib/device-log/ios-performance-log.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.IOSPerformanceLog = void 0;

require("source-map-support/register");

var _logger = _interopRequireDefault(require("./logger"));

var _lodash = _interopRequireDefault(require("lodash"));

const MAX_EVENTS = 5000;

class IOSPerformanceLog {
  constructor(remoteDebugger, maxEvents = MAX_EVENTS) {
    this.remoteDebugger = remoteDebugger;
    this.maxEvents = parseInt(maxEvents, 10);
    this.timelineEvents = [];
  }

  async startCapture() {
    _logger.default.debug('Starting performance (Timeline) log capture');

    this.timelineEvents = [];
    return await this.remoteDebugger.startTimeline(this.onTimelineEvent.bind(this));
  }

  async stopCapture() {
    _logger.default.debug('Stopping performance (Timeline) log capture');

    return await this.remoteDebugger.stopTimeline();
  }

  onTimelineEvent(event) {
    _logger.default.debug(`Received Timeline event: ${_lodash.default.truncate(JSON.stringify(event))}`);

    this.timelineEvents.push(event);

    if (this.timelineEvents.length > this.maxEvents) {
      let removedEvent = this.timelineEvents.shift();

      _logger.default.warn(`Too many Timeline events, removing earliest: ${_lodash.default.truncate(JSON.stringify(removedEvent))}`);
    }
  }

  async getLogs() {
    let events = this.timelineEvents;

    _logger.default.debug('Flushing Timeline events');

    this.timelineEvents = [];
    return events;
  }

  async getAllLogs() {
    return this.getLogs();
  }

}

exports.IOSPerformanceLog = IOSPerformanceLog;
var _default = IOSPerformanceLog;
exports.default = _default;require('source-map-support').install();
