// appium-ios-driver/build/lib/iwdp.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.IWDP = void 0;

require("source-map-support/register");

var _teen_process = require("teen_process");

var _events = _interopRequireDefault(require("events"));

var _logger = _interopRequireDefault(require("./logger"));

var _requestPromise = _interopRequireDefault(require("request-promise"));

var _asyncbox = require("asyncbox");

var _appiumSupport = require("appium-support");

const IWDP_CMD = 'ios_webkit_debug_proxy';
const MAX_RETRIES = 10;

const iwdpLogger = _appiumSupport.logger.getLogger('IWDP');

class IWDP extends _events.default {
  constructor(opts = {}) {
    super();
    this.udid = opts.udid || null;
    this.logStdout = !!opts.logStdout;
    this.attempts = 0;
    this.port = opts.webkitDebugProxyPort || 27753;
    this.process = this.createIWDPProcess();
    this.endpoint = `http://localhost:${this.port}`;
  }

  createIWDPProcess() {
    const process = new _teen_process.SubProcess(IWDP_CMD, ['-c', `${this.udid}:${this.port}`, '-d']);
    process.on('exit', () => this.onExit());
    process.on('lines-stderr', iwdpLogger.error);

    if (this.logStdout) {
      process.on('lines-stdout', iwdpLogger.debug);
    }

    return process;
  }

  async onExit() {
    if (!this.exitRequested) {
      this.process = this.createIWDPProcess();
      await this.start();
    }
  }

  async start() {
    if (++this.attempts > MAX_RETRIES) {
      return this.emit('error', new Error(`Failed to start IWDP server. Max retry attempts ${MAX_RETRIES} reached`));
    }

    this.exitRequested = false;

    if (!(await this.isSupported())) {
      _logger.default.errorAndThrow(`'ios_webkit_debug_proxy' not installed on this machine. Try 'brew install ios-webkit-debug-proxy`);
    }

    _logger.default.debug(`Starting ios_webkit_debug_proxy at port ${this.port} on device ${this.udid}`);

    await this.process.start(0);

    try {
      await (0, _asyncbox.retryInterval)(20, 500, async () => await (0, _requestPromise.default)(this.endpoint));
      this.emit('start');
    } catch (ign) {
      try {
        await this.process.stop();
      } catch (ign2) {}

      _logger.default.errorAndThrow(`Timed out waiting for ios_webkit_debug_proxy to open`);
    }
  }

  async stop() {
    this.attempts = 0;
    this.exitRequested = true;
    return await this.process.stop();
  }

  async isSupported() {
    if (typeof this.supported !== 'undefined') {
      return this.supported;
    }

    try {
      await _appiumSupport.fs.which(IWDP_CMD);
      this.supported = true;
    } catch (e) {
      this.supported = false;
    }

    return this.supported;
  }

}

exports.IWDP = IWDP;
var _default = IWDP;
exports.default = _default;require('source-map-support').install();
