// appium-ios-driver/build/lib/commands/alert.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

require("source-map-support/register");

var _appiumSupport = require("appium-support");

var _appiumBaseDriver = require("appium-base-driver");

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;

function handleError(err) {
  if (err.message && err.message.match(/not open/)) {
    throw new _appiumBaseDriver.errors.NoAlertOpenError();
  } else {
    throw err;
  }
}

commands.getAlertText = async function getAlertText() {
  try {
    let ret = await this.uiAutoClient.sendCommand('au.getAlertText()');
    return ret;
  } catch (err) {
    handleError(err);
  }
};

commands.setAlertText = async function setAlertText(text) {
  try {
    text = _appiumSupport.util.escapeSpecialChars(text, "'");
    await this.uiAutoClient.sendCommand(`au.setAlertText('${text}')`);
  } catch (err) {
    handleError(err);
  }
};

commands.postAcceptAlert = async function postAcceptAlert() {
  try {
    await this.uiAutoClient.sendCommand('au.acceptAlert()');
  } catch (err) {
    handleError(err);
  }
};

commands.postDismissAlert = async function postDismissAlert() {
  try {
    await this.uiAutoClient.sendCommand('au.dismissAlert()');
  } catch (err) {
    handleError(err);
  }
};

Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
