// appium-ios-driver/build/lib/commands/device.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

require("source-map-support/register");

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;

commands.getDevicePixelRatio = async function getDevicePixelRatio() {
  const screenHeight = (await this.getWindowSize()).height;
  const screenHeightOfPlusDevices = 736;
  return screenHeight === screenHeightOfPlusDevices ? 3.0 : 2.0;
};

commands.getViewportRect = async function getViewportRect() {
  const scale = await this.getDevicePixelRatio();
  const statusBarHeight = Math.round((await this.getStatusBarHeight()) * scale);
  const windowSize = await this.getWindowSize();
  return {
    left: 0,
    top: statusBarHeight,
    width: windowSize.width * scale,
    height: windowSize.height * scale - statusBarHeight
  };
};

commands.getStatusBarHeight = async function getStatusBarHeight() {
  const command = 'UIATarget.localTarget().frontMostApp().statusBar().rect().size.height;';
  return await this.uiAutoClient.sendCommand(command);
};

Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
