// appium-remote-debugger/build/lib/rpc-client.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _logger = _interopRequireDefault(require("./logger"));

var _remoteMessages = _interopRequireDefault(require("./remote-messages"));

var _asyncbox = require("asyncbox");

var _lodash = _interopRequireDefault(require("lodash"));

const WAIT_FOR_TARGET_RETRIES = 10;
const WAIT_FOR_TARGET_INTERVAL = 1000;

class RpcClient {
  constructor(opts = {}) {
    this._targets = [];
    this._shouldCheckForTarget = !!opts.shouldCheckForTarget;
  }

  setCommunicationProtocol(isTargetBased = false) {
    _logger.default.warn(`Setting communication protocol: using ${isTargetBased ? 'Target-based' : 'full Web Inspector protocol'} communication`);

    this.isTargetBased = isTargetBased;

    if (!this.remoteMessages) {
      this.remoteMessages = new _remoteMessages.default(isTargetBased);
    } else {
      this.remoteMessages.setCommunicationProtocol(isTargetBased);
    }
  }

  async waitForTarget() {
    if (!this.shouldCheckForTarget || !this.isTargetBased) {
      return;
    }

    await (0, _asyncbox.retryInterval)(WAIT_FOR_TARGET_RETRIES, WAIT_FOR_TARGET_INTERVAL, () => {
      if (_lodash.default.isEmpty(this.targets)) {
        throw new Error('No targets found, unable to communicate with device');
      }
    });
  }

  async send(command, opts = {}) {
    try {
      await this.waitForTarget();
      return await this.sendMessage(command, opts);
    } catch (err) {
      if (err.message.includes(`'Target' domain was not found`)) {
        this.setCommunicationProtocol(false);
        return await this.sendMessage(command, opts);
      } else if (err.message.includes(`domain was not found`)) {
        this.setCommunicationProtocol(true);
        await this.waitForTarget();
        return await this.sendMessage(command, opts);
      }

      throw err;
    }
  }

  async sendMessage() {
    throw new Error(`Sub-classes need to implement a 'sendMessage' function`);
  }

  addTarget(targetInfo) {
    if (_lodash.default.isUndefined(targetInfo) || _lodash.default.isUndefined(targetInfo.targetId)) {
      _logger.default.debug(`Received 'targetCreated' event with no target. Skipping`);

      return;
    }

    _logger.default.debug(`Target created: ${JSON.stringify(targetInfo)}`);

    if (!this.targets.includes(targetInfo.targetId)) {
      this.targets.push(targetInfo.targetId);
    }
  }

  removeTarget(targetInfo) {
    if (_lodash.default.isUndefined(targetInfo) || _lodash.default.isUndefined(targetInfo.targetId)) {
      _logger.default.debug(`Received 'taretDestroyed' event with no target. Skipping`);

      return;
    }

    _logger.default.debug(`Target destroyed: ${JSON.stringify(targetInfo)}`);

    _lodash.default.pull(this.targets, targetInfo.targetId);
  }

  get targets() {
    this._targets = this._targets || [];
    return this._targets;
  }

  get target() {
    if (_lodash.default.isEmpty(this.targets)) {
      throw new Error('No targets found, unable to communicate with device');
    }

    return _lodash.default.first(this.targets);
  }

  get shouldCheckForTarget() {
    return this._shouldCheckForTarget;
  }

  set shouldCheckForTarget(shouldCheckForTarget) {
    this._shouldCheckForTarget = !!shouldCheckForTarget;
  }

}

exports.default = RpcClient;require('source-map-support').install();
