// appium-android-driver/build/lib/commands/alert.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

require("source-map-support/register");

var _appiumBaseDriver = require("appium-base-driver");

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;

commands.getAlertText = async function getAlertText() {
  throw new _appiumBaseDriver.errors.NotYetImplementedError();
};

commands.setAlertText = async function setAlertText() {
  throw new _appiumBaseDriver.errors.NotYetImplementedError();
};

commands.postAcceptAlert = async function postAcceptAlert() {
  throw new _appiumBaseDriver.errors.NotYetImplementedError();
};

commands.postDismissAlert = async function postDismissAlert() {
  throw new _appiumBaseDriver.errors.NotYetImplementedError();
};

Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
