// appium-xcuitest-driver/build/lib/device-log/ios-crash-log.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.IOSCrashLog = void 0;

require("source-map-support/register");

var _appiumIosDriver = require("appium-ios-driver");

var _appiumSupport = require("appium-support");

var _bluebird = _interopRequireDefault(require("bluebird"));

var _logger = _interopRequireDefault(require("../logger"));

var _teen_process = require("teen_process");

var _path = _interopRequireDefault(require("path"));

class IOSCrashLog extends _appiumIosDriver.IOSCrashLog {
  constructor(opts = {}) {
    super(opts.udid ? _path.default.resolve(process.env.HOME, 'Library', 'Logs', 'CrashReporter', 'MobileDevice') : _path.default.resolve(process.env.HOME, 'Library', 'Logs', 'DiagnosticReports'));
    this.udid = opts.udid;
    this.phoneName = null;
    this.sim = opts.sim;
  }

  async getCrashes() {
    let crashLogsRoot = this.logDir;

    if (this.udid) {
      if (!this.phoneName) {
        try {
          const {
            stdout
          } = await (0, _teen_process.exec)('idevicename', ['-u', this.udid]);
          this.phoneName = stdout.trim();
        } catch (e) {
          _logger.default.warn(`Cannot get the name of the crashes folder for the device with udid '${this.udid}'. ` + `Original error: ${e.message}`);

          return [];
        }
      }

      if (this.phoneName) {
        crashLogsRoot = _path.default.resolve(crashLogsRoot, this.phoneName);
      }
    }

    if (!(await _appiumSupport.fs.exists(crashLogsRoot))) {
      _logger.default.debug(`Crash reports root '${crashLogsRoot}' does not exist. Got nothing to gather.`);

      return [];
    }

    const foundFiles = await _appiumSupport.fs.glob(`${crashLogsRoot}/**/*.crash`);

    if (this.udid) {
      return foundFiles;
    }

    return await _bluebird.default.filter(foundFiles, async x => {
      try {
        const content = await _appiumSupport.fs.readFile(x, 'utf8');
        return content.toUpperCase().includes(this.sim.udid.toUpperCase());
      } catch (err) {
        return false;
      }
    });
  }

  async filesToJSON(paths) {
    return await _bluebird.default.map(paths, async fullPath => {
      const stat = await _appiumSupport.fs.stat(fullPath);
      return {
        timestamp: stat.ctime.getTime(),
        level: 'ALL',
        message: await _appiumSupport.fs.readFile(fullPath, 'utf8')
      };
    });
  }

}

exports.IOSCrashLog = IOSCrashLog;
var _default = IOSCrashLog;
exports.default = _default;require('source-map-support').install();
