// appium-mac-driver/build/lib/driver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.MacDriver = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _appiumBaseDriver = require("appium-base-driver");

var _appiumSupport = require("appium-support");

var _appiumForMac = require("./appium-for-mac");

var _logger = _interopRequireDefault(require("./logger"));

class MacDriver extends _appiumBaseDriver.BaseDriver {
  constructor(opts = {}, shouldValidateCaps = true) {
    super(opts, shouldValidateCaps);
    this.jwpProxyActive = false;
    this.opts.address = opts.address || _appiumForMac.DEFAULT_A4M_HOST;
  }

  createSession(...args) {
    var _this = this,
        _superprop_callCreateSession = (..._args) => super.createSession(..._args);

    return (0, _asyncToGenerator2.default)(function* () {
      if (!_appiumSupport.system.isMac()) {
        throw new Error("AppiumForMac tests only run on the Mac");
      }

      try {
        let _ref = yield _superprop_callCreateSession(...args),
            _ref2 = (0, _slicedToArray2.default)(_ref, 2),
            sessionId = _ref2[0],
            caps = _ref2[1];

        yield _this.startAppiumForMacSession();

        if (caps.app) {
          _logger.default.info(`Automatically navigating to app '${caps.app}'`);

          yield _this.a4mDriver.sendCommand('/url', 'POST', {
            url: caps.app
          });
        }

        return [sessionId, caps];
      } catch (e) {
        yield _this.deleteSession();
        throw e;
      }
    })();
  }

  startAppiumForMacSession() {
    var _this2 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _this2.a4mDriver = new _appiumForMac.AppiumForMac();
      yield _this2.a4mDriver.start();
      yield _this2.a4mDriver.startSession(_this2.caps);
      _this2.proxyReqRes = _this2.a4mDriver.proxyReqRes.bind(_this2.a4mDriver);
      _this2.jwpProxyActive = true;
    })();
  }

  deleteSession() {
    var _this3 = this,
        _superprop_callDeleteSession = (..._args2) => super.deleteSession(..._args2);

    return (0, _asyncToGenerator2.default)(function* () {
      _logger.default.debug('Deleting AppiumForMac session');

      if (_this3.a4mDriver && _this3.jwpProxyActive) {
        yield _this3.a4mDriver.deleteSession();
        yield _this3.a4mDriver.stop();
        _this3.a4mDriver = null;
      }

      _this3.jwpProxyActive = false;
      yield _superprop_callDeleteSession();
    })();
  }

  proxyActive() {
    return true;
  }

  canProxy() {
    return true;
  }

  get driverData() {
    return {
      A4MPort: this.opts.port
    };
  }

}

exports.MacDriver = MacDriver;
var _default = MacDriver;
exports.default = _default;require('source-map-support').install();
