// appium-base-driver/build/lib/protocol/validators.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.validators = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

function isNumber(o) {
  return _lodash.default.isNumber(o) || !_lodash.default.isNaN(parseInt(o, 10)) || !_lodash.default.isNaN(parseFloat(o));
}

function msValidator(ms) {
  if (!_lodash.default.isNumber(ms) || ms < 0) {
    throw new Error('Wait ms must be a number equal to 0 or greater');
  }
}

const validators = {
  setUrl: url => {
    if (!url || !url.match(/^([a-zA-Z0-9_+.-]+:\/\/)|(about:)|(data:)/)) {
      throw new Error('Url or Uri must start with <scheme>://');
    }
  },
  implicitWait: ms => {
    msValidator(ms);
  },
  asyncScriptTimeout: ms => {
    msValidator(ms);
  },
  clickCurrent: button => {
    if (!(isNumber(button) || _lodash.default.isUndefined(button)) || button < 0 || button > 2) {
      throw new Error('Click button must be 0, 1, or 2');
    }
  },
  setNetworkConnection: type => {
    if (!isNumber(type) || [0, 1, 2, 4, 6].indexOf(type) === -1) {
      throw new Error('Network type must be one of 0, 1, 2, 4, 6');
    }
  }
};
exports.validators = validators;require('source-map-support').install();
