// appium-fake-driver/build/lib/driver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FakeDriver = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _lodash = _interopRequireDefault(require("lodash"));

var _appiumBaseDriver = require("appium-base-driver");

var _fakeApp = require("./fake-app");

var _commands = _interopRequireDefault(require("./commands"));

class FakeDriver extends _appiumBaseDriver.BaseDriver {
  constructor() {
    super();
    this.appModel = null;
    this.curContext = 'NATIVE_APP';
    this.elMap = {};
    this.focusedElId = null;
    this.maxElId = 0;
    this.caps = {};
    this.desiredCapConstraints = {
      app: {
        presence: true,
        isString: true
      }
    };
  }

  createSession(desiredCaps, requiredCaps, capabilities, otherSessionData = []) {
    var _this = this,
        _superprop_callCreateSession = (..._args) => super.createSession(..._args);

    return (0, _asyncToGenerator2.default)(function* () {
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = otherSessionData[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          let d = _step.value;

          if (d.isUnique) {
            throw new _appiumBaseDriver.errors.SessionNotCreatedError("Cannot start session; another " + "unique session is in progress that requires all resources");
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return != null) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      let _ref = yield _superprop_callCreateSession(desiredCaps, requiredCaps, capabilities, otherSessionData),
          _ref2 = (0, _slicedToArray2.default)(_ref, 2),
          sessionId = _ref2[0],
          caps = _ref2[1];

      _this.appModel = new _fakeApp.FakeApp();

      if (_lodash.default.isArray(caps) === true && caps.length === 1) {
        caps = caps[0];
      }

      _this.caps = caps;
      yield _this.appModel.loadApp(caps.app);
      return [sessionId, caps];
    })();
  }

  get driverData() {
    return {
      isUnique: !!this.caps.uniqueApp
    };
  }

}

exports.FakeDriver = FakeDriver;
var _iteratorNormalCompletion2 = true;
var _didIteratorError2 = false;
var _iteratorError2 = undefined;

try {
  for (var _iterator2 = _lodash.default.toPairs(_commands.default)[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
    let _step2$value = (0, _slicedToArray2.default)(_step2.value, 2),
        cmd = _step2$value[0],
        fn = _step2$value[1];

    FakeDriver.prototype[cmd] = fn;
  }
} catch (err) {
  _didIteratorError2 = true;
  _iteratorError2 = err;
} finally {
  try {
    if (!_iteratorNormalCompletion2 && _iterator2.return != null) {
      _iterator2.return();
    }
  } finally {
    if (_didIteratorError2) {
      throw _iteratorError2;
    }
  }
}require('source-map-support').install();
