// appium-fake-driver/build/lib/fake-element.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FakeElement = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _lodash = _interopRequireDefault(require("lodash"));

var _xmldom = _interopRequireDefault(require("xmldom"));

class FakeElement {
  constructor(xmlNode, app) {
    this.app = app;
    this.node = xmlNode;
    this.nodeAttrs = {};
    this.type = this.node.tagName;
    this.attrs = {};
    this.css = {};
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = _lodash.default.values(this.node.attributes)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        let _step$value = _step.value,
            name = _step$value.name,
            value = _step$value.value;
        this.nodeAttrs[name] = value;
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return != null) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    this.parseCss();
  }

  parseCss() {
    if (this.nodeAttrs.style) {
      let segments = this.nodeAttrs.style.split(";");
      var _iteratorNormalCompletion2 = true;
      var _didIteratorError2 = false;
      var _iteratorError2 = undefined;

      try {
        for (var _iterator2 = segments[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
          let s = _step2.value;

          let _s$split = s.split(":"),
              _s$split2 = (0, _slicedToArray2.default)(_s$split, 2),
              prop = _s$split2[0],
              val = _s$split2[1];

          prop = prop.trim();
          val = val.trim();
          this.css[prop] = val;
        }
      } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion2 && _iterator2.return != null) {
            _iterator2.return();
          }
        } finally {
          if (_didIteratorError2) {
            throw _iteratorError2;
          }
        }
      }
    }
  }

  get tagName() {
    return this.node.tagName;
  }

  setAttr(k, v) {
    this.attrs[k] = v;
  }

  getAttr(k) {
    return this.attrs[k] || "";
  }

  isVisible() {
    return this.nodeAttrs.visible !== "false";
  }

  isEnabled() {
    return this.nodeAttrs.enabled !== "false";
  }

  isSelected() {
    return this.nodeAttrs.selected === "true";
  }

  getLocation() {
    return {
      x: parseFloat(this.nodeAttrs.left || 0),
      y: parseFloat(this.nodeAttrs.top || 0)
    };
  }

  getSize() {
    return {
      width: parseFloat(this.nodeAttrs.width || 0),
      height: parseFloat(this.nodeAttrs.height || 0)
    };
  }

  click() {
    let curClicks = this.getAttr('clicks') || 0;
    this.setAttr('clicks', curClicks + 1);
    let alertId = this.nodeAttrs.showAlert;

    if (alertId) {
      this.app.showAlert(alertId);
    }
  }

  equals(other) {
    return this.node === other.node;
  }

  hasPrompt() {
    return this.nodeAttrs.hasPrompt === "true";
  }

  getCss(prop) {
    if (_lodash.default.has(this.css, prop)) {
      return this.css[prop];
    }

    return null;
  }

  get xmlFragment() {
    let frag = new _xmldom.default.XMLSerializer().serializeToString(this.node);
    return new _xmldom.default.DOMParser().parseFromString(frag, "application/xml");
  }

}

exports.FakeElement = FakeElement;require('source-map-support').install();
